### ISSUE TITLE
Linebreaking support

### ISSUE THREAD

**Issue by fred-wang** [mathml-refresh/mathml#112]
cc @bfgeek

The MathML core spec now defines all the min-content / max-content values, however these two are equal and linebreaking is supposed to never happens.

I believe linebreaking could potentially happen in:

- Inline equations. However, I think defining CSS fragments will be a bit complicate.

- mrow-like element from display equations. We have an experiment like this in chromium. Currently that depends on the parent's width however, we could do something similar if width is specified on the element (#45).

- mtable (I guess that's already defined by CSS). Maybe we want to prevent it for now.

MathML 3 rules for linebreaking are quite complex, maybe we should have a simple version first and refine it later or get it improved by polyfills when the CSS Layout API is ready.

Just opening this so that it can be referenced from the spec.

**Comment by fred-wang** [comment]
We need to investigate a bit this, but I imaging we could introduce a math-wrap property (or rely on an existing CSS one) in the future, which would default to nowrap if we are concerned about backward-compatibility change.

**Comment by fred-wang** [comment]
Gecko disables linebreaking in table cell:
https://dxr.mozilla.org/mozilla-central/source/layout/mathml/mathml.css#137

WebKit disables linebreaking in foreign content: 
https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/css/mathml.css#L134

I guess we could set white-space to nowrap by default on MathML elements to prevent any backward-compatibility issue.

**Comment by fred-wang** [comment]
It would probably interesting to start experiment line breaking with the CSS Layout API ( https://drafts.css-houdini.org/css-layout-api/ ) maybe starting with basic tests and then trying write a mrow-like layout. 

In any case, I think #130 should be resolved first.

**Comment by ronkok** [comment]
Line breaks are more important in the mobile-screen world than they were years ago. I think this should get a high priority.

__The TeXbook__, page 173 states that a "A formula will be broken only after a relation symbol like $=$ or $<$ or $\rightarrow$, or after a binary operation symbol like $+$ or $-$ or $\times$, where the relation or binary operation is on the ``outer level'' of the formula (i.e., not enclosed in {...} and not part of an \over construction)."

KaTeX does its best to emulate the TeXbook rule. For my own work, this issue the is single thing that will cause me to use use KaTeX HTML rather than MathML.

**Comment by NSoiffer** [comment]
The rules that TeX has for linebreaking only apply to _inline_ math. TeX categorizes symbols into a small number of categories (to fit in four bits because efficiency was critical at the time of its design). This means that there are broad generalizations that work a lot of the time but not always. E.g., `(a+b)⋅(a-b)` might break at the `+` or `-`, which would be a poor place to break.

Generally, in linebreaking you want to look at the expression tree and break as close to the root as possible while still filling as much of the line as possible. Typically, relational operators will be at the root of a tree, then lower precedence operators like + and -, then higher precedence operators like ⋅, ⨯, and /. Well-structured MathML (which is unfortunately not very common) has mrows that align with the expression tree so knowing where is a good spot to break is easier with well structured MathML. [MathML's operator dictionary](https://www.w3.org/TR/MathML3/appendixc.html) gives priorities of operators that can be used for parsing and linebreaking, and serve as a guide for spacing also (higher priorities have less space around them in general). MathML 3 lists [a potential linebreaking algorithm](https://www.w3.org/TR/MathML/chapter3.html#presm.lbalgorithm) that takes time proportional to the number of token elements times the number of lines, so it is relatively quick and does a pretty good job. A more complicated version that looks at the whole expression vs a single line would mimic TeX's paragraph linebreaking rules. I would like to see the simpler algorithm become part of the core spec, but I understand we need to have priorities and this might add several weeks of implementation and spec time if cleanly doable at all with the current state of CSS.

In response to earlier comments:
1. Linebreaking inside of tables/matrices is complicated if you want to do a good job. That's because if the width of the table cell is computed automatically, you really want to do that knowing how good/bad that width is for linebreaking the expression. If there is a fixed column width, then it is no more complicated than normal linebreaking. In my former job, we had a publisher that had two column layout and put math inside of tables in each column, so they were very concerned about good linebreaking. The point of this comment is that allowing linebreaks in math in tables will address real needs of publishers. On phones, those issues will be there for any math in a table.

2. In addition to linebreaks, for display math, one needs to deal with indentation. As with linebreaks, indentation levels reflect the expression tree and help readers understand what is grouped with what.

3.  Linebreaking does require another pass over the layout once sizes have been determined, but it is not rocket science to figure out good linebreaks, at least conceptually. Whether it fits in with the CSS layout model is (I think) the main question. The [CSS Layout API](https://drafts.css-houdini.org/css-layout-api) mentioned earlier seems promising with indentation maybe done by left-padding each line.

4. Any polyfill that did linebreaking would cause reflow, which would be bad. It would have to come after layout is done. Potentially a polyfill could be written that creates well-structured mrows so that the browser implementation has an easier time doing linebreaking.

5. MathML 3 has a number of manual linebreaking and indenting options that can set on `mo`. Maybe the first step is for the core spec to specify those. That would at least allow an author to get some linebreaking/indentation to happen so that (for example) an expression with multiple `=` signs can be broken at the `=`s and aligned.

**Comment by fred-wang** [comment]
I think there are two important points in Neil's reply:

* We first need to focus on properly implementing a CSS-compatible and interoperable math rendering in all browsers, which is what we are doing with MathML Core. This must be incremental and already starting to introduce a lot of advanced features before the first step is done is counter-productive if we want to keep the current support we've gotten from browser vendors or CSS WG people. Everybody definitely agree that linebreaking is an important property of text/CSS/math layout, though.

* The idea with MathML Core is that you can just use normal CSS/JS technologies as for other HTML elements. So if KaTeX or other polyfills are able to do linebreaking with HTML elements they could just follow similar approach with MathML elements. With the current technologies available (e.g. without CSS layout API) that would be bad for performance because it would require forcing relayout but it's probably not that bad compared to what these polyfills are doing right now (which requires to do the full math layout themselves). So I think it could be an interesting experiment to try in the short term.

**Comment by ronkok** [comment]
> This must be incremental

Agreed and acknowledged. The work being done is excellent.

> MathML 3 has a number of manual linebreaking and indenting options that can set on `mo`. 

Yes, MathML 3 contemplates an attribute of `linebreakstyle` on a `<mo>`. It would be great if this were to be specified and actually implemented, unlike in current FireFox.

> The idea with MathML Core is that you can just use normal CSS/JS technologies as for other HTML elements.

If I understand that statement correctly, then one could apply an inline style of `display: inline-block` to a top-level `<mo>` and it would act like just like a `<mo>` with a `linebreakstyle="before"` attribute. That would also be terrific and would be all that I ask.

Do I understand that statement correctly?

**Comment by ronkok** [comment]
> So if KaTeX or other polyfills are able to do linebreaking with HTML elements they could just follow similar approach with MathML elements.

A similar approach would break the top level into multiple `mrow`s, with each break occurring at a binary or relational operator. That would map `a + b = d` into:

    <mrow><mi>a</mi><mo>+</mo></mrow>
    <mrow><mi>b</mi><mo>=</mo></mrow>
    <mrow><mi>c</mi></mrow>

This approach would create automatic line breaks in the __TeXbook__ locations. It works, at some cost to the semantics. 

As suggested, KaTeX could implement this method. It is very similar to what is now done in HTML. If the method in the previous statement will not work, the method in this comment is probably what we will do. Let me know where we stand.

**Comment by fred-wang** [comment]
> > So if KaTeX or other polyfills are able to do linebreaking with HTML elements they could just follow similar approach with MathML elements.
> 
> A similar approach would break the top level into multiple `mrow`s, with each break occurring at a binary or relational operator. That would map `a + b = d` into:

This is the short term approach I was thinking about. You can use `getBoundingClientRect()` to know position and size after layout in order to apply line breaking depending on the screen size. If semantics is a problem, note that you can put these splited MathML pieces into a shadow tree so that the original MathML DOM is still available.

**Comment by ronkok** [comment]
@fred-wang Thank you for the quick response. That clears up the picture considerably.

**Comment by fred-wang** [comment]
@ronkok No problem. Additionally, note that you can use https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver to watch when the width of the container of the `<math>` tag changes in order to update linebreaking(e.g. when the user resize the window.

**Comment by ronkok** [comment]
KaTeX avoids, so far, any reliance on the browser for runtime information. It generates code that works whether generated client-side or server-side. So I think we're stuck with the multiple `<mrow>` approach.

**Comment by fred-wang** [comment]
consensus from 2020/06/23: postpone to a future version
